arch/x86/core/cache.c

```diff
@@ -33,20 +33,22 @@ static void arch_dcache_flush(void *start_addr, size_t size)
 {
 	size_t line_size = sys_cache_data_line_size_get();
 	uintptr_t start = (uintptr_t)start_addr;
-	uintptr_t end;
+	uintptr_t end = start + size;
 
 	if (line_size == 0U) {
 		return;
 	}
 
-	size = ROUND_UP(size, line_size);
-	end = start + size;
+	end = ROUND_UP(end, line_size);
 
 	for (; start < end; start += line_size) {
-		__asm__ volatile("clflush %0;\n\t" :  : "m"(start));
+		__asm__ volatile("clflush %0;\n\t" :
+				"+m"(*(volatile char *)start));
 	}
 
+#if defined(CONFIG_X86_MFENCE_INSTRUCTION_SUPPORTED)
 	__asm__ volatile("mfence;\n\t");
+#endif
```

Comment:
mfence is being used in `apic_tsc` driver, should it be checked there as well ?

Reply:
btw, GCC provides `__sync_synchronize()` that is a compiler and memory barrier.

Reply:
@ceolin - are these gating issues?

Reply:
@ceolin You raised a good point. I googled and looked Linux code, added memory barrier code without MFENCE.
tried __sync_synchronize(). it will add similar code with "lock" instruction.
I have no board supporting apic_tsc, so I'm afraid to change it's code.
apic_tsc seems to be a new x86 feature and co-exist with SSE2/MFENCE. Can my patch just ignore MFENCE in apic_tsc?

Reply:
Sounds good for me !